### ISSUE TITLE
KTOTrainer Memory Leakage 

### ISSUE THREAD

**Issue by Isaaclgz** [huggingface/trl#2268]
### System Info

I've been running some experiments on KTO using LoRA and noticed that there is a large disparity between the peak allocated and reserved memory, which I suspect is a memory leakage problem. 

To my understanding, when using LoRA with KTOTrainer, the target model is also used as the reference model for calculating the reward and approximate KL-divergence by [disabling the adapter weights ](https://github.com/huggingface/trl/blob/88be2c07e5024cc47e7777d23815602ae22c5f11/trl/trainer/kto_trainer.py#L467). I am unfamiliar with how gradients and optimizer states might differ when performing KTO compared to SFT, but intuitively I don't think they should be much different, if at all. And so, if we aren't loading a separate model as the reference model, the memory usage should not be that high. 

For reference, I am performing KTO with the following parameters, and using a variant of Llama-3-8B-Instruct (called 'aisingapore/llama3-8b-cpt-sea-lionv2.1-instruct'): 
```
training_args = KTOConfig(
            num_train_epochs=2,
            per_device_train_batch_size=4,
            remove_unused_columns=False,
            gradient_accumulation_steps=4,
            learning_rate=5e-7,
            evaluation_strategy="steps",
            beta=0.5,
            max_length=1000,
            gradient_checkpointing=True,
            bf16 = True, 
            save_total_limit=0,
            undesirable_weight=1.0,
            max_prompt_length=800,
            max_completion_length=200,
            report_to="none"
)
```
Tracking memory usage with `torch.cuda` utilities, for KTO I have: 
![image](https://github.com/user-attachments/assets/1509a27f-f3c9-4ed3-920d-0a075c1d24d9)

but for SFT on the same dataset with the same batch size, etc:
![image](https://github.com/user-attachments/assets/040dd622-0731-4c77-b54b-4e82a1328672)

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

def torch_print_gpu_memory():
    print(f"Memory allocated: {torch.cuda.memory_allocated() / 1024 ** 3:.4f} GB")
    print(f"Memory reserved: {torch.cuda.memory_reserved() / 1024 ** 3:.4f} GB")
    print(f"Peak memory allocated: {torch.cuda.max_memory_allocated() / 1024**3:.4f} GB")
    print(f"Peak memory reserved: {torch.cuda.max_memory_reserved() / 1024**3:.4f} GB")

print('Before loading base model')
torch_print_gpu_memory()
print('\n----------\n')

model = AutoModelForCausalLM.from_pretrained(
    'aisingapore/llama3-8b-cpt-sea-lionv2.1-instruct', 
    trust_remote_code=True, 
    device_map = 'auto',
    torch_dtype = torch.bfloat16,
    attn_implementation="flash_attention_2"
) 

print('\n----------\n')
print('After loading base model')
torch_print_gpu_memory()
print('\n----------\n')

lora_config = LoraConfig(
    r=128,
    lora_alpha=128,
    lora_dropout=0.05,
    target_modules= 'all-linear',
)

training_args = KTOConfig(
            num_train_epochs=2,
            per_device_train_batch_size=4,
            remove_unused_columns=False,
            gradient_accumulation_steps=4,
            learning_rate=5e-7,
            evaluation_strategy="steps",
            beta=0.5,
            max_length=1000,
            gradient_checkpointing=True,
            bf16 = True, 
            save_total_limit=0,
            undesirable_weight=1.0,
            max_prompt_length=800,
            max_completion_length=200,
            report_to="none"
)

print('Before KTOTrainer Instantiation')
torch_print_gpu_memory()
print('\n----------\n')

trainer = KTOTrainer(
    model=model,
    args=training_args,
    tokenizer=tokenizer,
    train_dataset=dataset,
    peft_config=lora_config,
)

print('\n----------\n')
print('After KTOTrainer Instantiation')
torch_print_gpu_memory()
print('\n----------\n')

print('\n----------\n')
print('Before KTO Train')
torch_print_gpu_memory()
print('\n----------\n')

trainer.train()

print('\n----------\n')
print('After KTO Train')
torch_print_gpu_memory()
print('\n----------\n')

### Expected behavior

Peak memory reserved when using KTOTrainer for PEFT should be significantly closer to peak memory allocated.

**Comment by qgallouedec** [comment]
Thanks for reporting, please share your system info

**Comment by Isaaclgz** [comment]
> Thanks for reporting, please share your system info

Thanks for looking into this! 

System: 
Debian 11
Python 3.10
1xA100-80GB
Nvidia driver 550.90.07, CUDA 12.4
(running this on a GCP CE instance based on the c0-deeplearning-common-cu123-v20240922-debian-11-py310 image)

Env:
torch==2.4.0
transformers==4.44.0
trl==0.11.3
flash-attn==2.6.3
accelerate==1.0.1

**Comment by chenyang399** [comment]
is there any chance that we can run KTO script with 24G GPU

**Comment by albertvillanova** [comment]
Hey @Isaaclgz, sorry for the very slow reply — we're catching up on the backlog. 

What you're seeing looks like expected behavior rather than a leak: a large gap between peak *reserved* and peak *allocated* is caching-allocator fragmentation, not memory that's being lost. KTO runs several differently-sized forward passes per step (the target completion, the shuffled KL completion, and the adapter-disabled reference forwards for LoRA), and those variable-length allocations fragment the allocator and inflate *reserved* — but the gradient-bearing work is still just one forward+backward on the target (the KL and reference passes are all under torch.no_grad()), so the trainable footprint is close to SFT. Two things help directly: precompute_ref_log_probs=True computes the reference log-probs once up front so the per-step reference forwards disappear (see https://github.com/huggingface/trl/blob/988b5a0dc1b3c3579115ceafd80f1f01915e9ab7/trl/trainer/kto_config.py#L61-L68), and PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True shrinks the reserved/allocated gap (activation_offloading=True is also available now). The trainer has also been substantially rewritten since the 0.11.3 you ran. 

Closing as expected behavior, but please re-open with fresh numbers on a current TRL if you still see *allocated* memory growing unboundedly across steps — that would be a genuine leak.
